src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java

```diff
@@ -0,0 +1,388 @@
+    }
+
+    @Override
+    public void decorateQualityGateStatus(AnalysisDetails analysis) {
+        LOGGER.info("starting to analyze with " + analysis.toString());
+        String revision = analysis.getCommitSha();
+
+        try {
+            Configuration configuration = configurationRepository.getConfiguration();
+            final String hostURL = getMandatoryProperty(PULLREQUEST_GITLAB_URL, configuration);
+            final String apiToken = getMandatoryProperty(PULLREQUEST_GITLAB_TOKEN, configuration);
+            final String repositorySlug = getMandatoryProperty(PULLREQUEST_GITLAB_REPOSITORY_SLUG, configuration);
+            final String pullRequestId = analysis.getBranchName();
+
+            final boolean summaryCommentEnabled = Boolean.parseBoolean(getMandatoryProperty(PULL_REQUEST_COMMENT_SUMMARY_ENABLED, configuration));
+            final boolean fileCommentEnabled = Boolean.parseBoolean(getMandatoryProperty(PULL_REQUEST_FILE_COMMENT_ENABLED, configuration));
+            final boolean deleteCommentsEnabled = Boolean.parseBoolean(getMandatoryProperty(PULL_REQUEST_DELETE_COMMENTS_ENABLED, configuration));
+
+            final String restURL = String.format("%s/api/v4", hostURL);
+            final String userURL = restURL + "/user";
+            final String projectURL = restURL + String.format("/projects/%s", URLEncoder.encode(repositorySlug, StandardCharsets.UTF_8.name()));
+            final String statusUrl = projectURL + String.format("/statuses/%s", revision);
+            final String commitURL = projectURL + String.format("/repository/commits/%s", revision);
+            final String commitCommentUrl = commitURL + "/comments";
+            final String mergeRequestURl = projectURL + String.format("/merge_requests/%s", pullRequestId);
+            final String prCommitsURL = mergeRequestURl + "/commits";
+
+
+            LOGGER.info(String.format("Status url is: %s ", statusUrl));
+            LOGGER.info(String.format("Commit comment url is: %s ", commitCommentUrl));
+            LOGGER.info(String.format("PR commits url is: %s ", prCommitsURL));
+            LOGGER.info(String.format("User url is: %s ", userURL));
+
+            Map<String, String> headers = new HashMap<>();
+            headers.put("PRIVATE-TOKEN", apiToken);
+            headers.put("Accept", "application/json");
+
+            User user=getUser(userURL, headers);
+            LOGGER.info(String.format("Using user: %s ", user.getUsername()));
+
+            List<Commit> commits = getMRCommits(prCommitsURL, headers, deleteCommentsEnabled);
+            LOGGER.info(String.format("Commits in MR: %s ", commits.stream().map(Commit::getId).collect(Collectors.joining(", "))));
+            List<CommitDiscussion> commitDiscussions = new ArrayList<>();
+            for (Commit commit : commits) {
+                getCommitDiscussions(projectURL + String.format("/repository/commits/%s/discussions", commit.getId()), headers, deleteCommentsEnabled)
```

Comment:
what's the difference between creating discussions on MR's commits VS directly on MR?

there's an API for creating discussions or MR directly: https://docs.gitlab.com/ee/api/discussions.html#create-new-merge-request-thread

Reply:
I am unsure about the difference between the 2 APIs. The one I'm using also creates a discussion, that can be replied to.

I will also try the MR discussion API out, to see if that has any differences. But I want to rebase, cleanup and test the current code before I do that.

Reply:
cool :+1: 

what I would check is if discussion on commit shows up as thread in MR:

![image](https://user-images.githubusercontent.com/20919/71890273-354d0a80-3144-11ea-8ce3-8546df6f6b0d.png)

and if you turn on:
![image](https://user-images.githubusercontent.com/20919/71890353-5f9ec800-3144-11ea-97de-8a1ba881d52f.png)

then you should see:
![image](https://user-images.githubusercontent.com/20919/71890330-53b30600-3144-11ea-8f82-5bc5f8bbd9ea.png)

Reply:
I don't see that with the commit comments.

I will play around with the MR discussions.

Reply:
I have played around with MR discussions. code and result can be seen here: https://gitlab.com/tisoft/sonarqube-community-branch-plugin-gitlab-test/merge_requests/1

Resolving of discussions is currently not done by the plugin, it will delete all discussions and recreate them at each run.

What do you think?

Reply:
nice, looks awesome, well done! :100: :muscle: 

may I suggest adding a link to Sonar issue and/or Sonar rule? :angel:

I would also try creating discussions first, and then creating MR comment with summary, so that would also be at the end (before user comments, of course). What do you think?

not sure if it's something on my end (tried also incognito), but images aren't loading for me:
![image](https://user-images.githubusercontent.com/20919/71916552-7530e380-317e-11ea-988c-ed385b424ba0.png)

other than that, looks really nice, thanks for trying this out! :blush:

Reply:
Personally I like having the summary on top, so that I have a quick overview and don't need to do endless scrolling before I can get to that.

Images are loaded from the github master branch of this repo, and since this is not yet merged into master, there are no images yet. The contents of the summary and issue comments is shared between all the decoration implementations, I don't want to change that as part as of this PR, but I agree that a link back to sonar would be helpful. Will do a followup PR as soon as everything else is settled.

![image](https://user-images.githubusercontent.com/139041/71917358-24ba8580-3180-11ea-9049-5ce3236b25b3.png)

If you click on the failed pipeline marker on the righrt and then on SonarQube, you are send to the sonar instance. (which will not work since its running on localhost on my machine :laughing:)